### ISSUE TITLE
[Bug] CrossThreadReduction Lowering Issue

### ISSUE THREAD

**Issue by zxybazh** [apache/tvm#12976]
When lowering certain cuda workloads, CrossThreadReduction may fail to lower to all reduce and creating racing condition.
[Script](https://gist.github.com/zxybazh/398b217108743af965503d3782a0bad1) available here to print out the lowered TIR.
Example TIR:
```python
@tvm.script.ir_module
class Module:
    @T.prim_func
    def main(p0: T.Buffer[(1, 7, 7, 512), "float32"], tensor: T.Buffer[(1, 1, 1, 512), "float32"]) -> None:
        # function attr dict
        T.func_attr({"tir.noalias": True, "global_symbol": "main"})
        # body
        with T.block("root"):
            T.reads()
            T.writes()
            T.block_attr({"meta_schedule.unroll_explicit":512})
            tensor_1_shared = T.alloc_buffer([1, 1, 1, 512], dtype="float32", scope="shared")
            for i0_i1_i2_i3_0_fused in T.thread_binding(4, thread="blockIdx.x"):
                for ax0, ax1, ax2, ax3, ax4_ax5_fused_0 in T.grid(1, 1, 1, 128, 1):
                    for ax4_ax5_fused_1 in T.thread_binding(128, thread="threadIdx.x"):
                        with T.block("tensor"):
                            T.where(ax4_ax5_fused_0 * 128 + ax4_ax5_fused_1 < 49)
                            ax0_1, ax1_1, ax2_1 = T.axis.remap("SSS", [ax0, ax1, ax2])
                            ax3_1 = T.axis.spatial(512, i0_i1_i2_i3_0_fused * 128 + ax3)
                            rv0 = T.axis.reduce(7, (ax4_ax5_fused_0 * 128 + ax4_ax5_fused_1) // 7)
                            rv1 = T.axis.reduce(7, (ax4_ax5_fused_0 * 128 + ax4_ax5_fused_1) % 7)
                            T.reads(p0[ax0_1, ax1_1 * 7 + rv0, ax2_1 * 7 + rv1, ax3_1])
                            T.writes(tensor_1_shared[ax0_1, ax1_1, ax2_1, ax3_1])
                            with T.init():
                                tensor_1_shared[ax0_1, ax1_1, ax2_1, ax3_1] = T.float32(0)
                            tensor_1_shared[ax0_1, ax1_1, ax2_1, ax3_1] = tensor_1_shared[ax0_1, ax1_1, ax2_1, ax3_1] + p0[ax0_1, ax1_1 * 7 + rv0, ax2_1 * 7 + rv1, ax3_1]
                for i3_1 in T.thread_binding(128, thread="threadIdx.x"):
                    with T.block("tensor_1"):
                        ax0 = T.axis.spatial(1, 0)
                        ax1 = T.axis.spatial(1, 0)
                        ax2 = T.axis.spatial(1, 0)
                        ax3 = T.axis.spatial(512, i0_i1_i2_i3_0_fused * 128 + i3_1)
                        T.reads(tensor_1_shared[ax0, ax1, ax2, ax3])
                        T.writes(tensor[ax0, ax1, ax2, ax3])
                        tensor[ax0, ax1, ax2, ax3] = tensor_1_shared[ax0, ax1, ax2, ax3] * T.float32(0.020408163265306121)
```

### Expected behavior

Cross thread reduction should be reduced to all reduce using `T.tvm_thread_allreduce`

### Actual behavior

Lowered TIR
```python
class Module:
    @T.prim_func
    def main(p0: T.Buffer[25088, "float32"], tensor: T.Buffer[512, "float32"]) -> None:
        # function attr dict
        T.func_attr({"tir.noalias": True, "global_symbol": "main"})
        # var definition
        threadIdx_x = T.env_thread("threadIdx.x")
        blockIdx_x = T.env_thread("blockIdx.x")
        T.preflattened_buffer(p0, [1, 7, 7, 512], dtype="float32", data=p0.data)
        T.preflattened_buffer(tensor, [1, 1, 1, 512], dtype="float32", data=tensor.data)
        # body
        T.launch_thread(blockIdx_x, 4)
        tensor_1_shared = T.allocate([128], "float32", "shared")
        tensor_1_shared_1 = T.buffer_decl([128], dtype="float32", data=tensor_1_shared, scope="shared")
        T.launch_thread(threadIdx_x, 128)
        for ax3 in T.serial(128):
            if threadIdx_x < 49:
                if threadIdx_x // 7 == 0 and threadIdx_x % 7 == 0:
                    tensor_1_shared_1[ax3] = T.float32(0)
                tensor_1_shared_1[ax3] = tensor_1_shared_1[ax3] + p0[threadIdx_x * 512 + blockIdx_x * 128 + ax3]
        tensor[blockIdx_x * 128 + threadIdx_x] = tensor_1_shared_1[threadIdx_x] * T.float32(0.020408163265306121)
```

### Environment

TVM Commit `8af43d3c11b13108eec0cea784961cf7b855c10f`, TIR code generated by running space generation for `nvidia/geforce-rtx-3070` target, space generation script available [here](https://gist.github.com/zxybazh/6aa9aab11c5622e759217aaaf040c248).

### Steps to reproduce

See script, simply do a `tvm.lower` can generate the current lowered TIR.

CC @junrushao @vinx13 

cc @Hzfengsy

**Comment by vinx13** [comment]
cc @MasterJH5574

**Comment by MasterJH5574** [comment]
Thanks for letting me know! Is this original PrimFunc hand-crafted or generated by a schedule?

**Comment by zxybazh** [comment]
Hi it's generated by a schedule, see script [here](https://gist.github.com/zxybazh/6aa9aab11c5622e759217aaaf040c248).

**Comment by MasterJH5574** [comment]
@zxybazh Thanks! Just want to know how much this bug blocks your current progress? As I’m very busying with other things this week, I might not have much time looking into it now.

**Comment by zxybazh** [comment]
Hi Hongyi, I can work around it right now so take your time. Thanks for paying close attention!

**Comment by zxybazh** [comment]
Hi @jinhongyii please let me know if you could reproduce this issue, thanks!

**Comment by zxybazh** [comment]
@MasterJH5574 Would you please take another look when you got time, thanks!

**Comment by junrushao** [comment]
@zxybazh would you like to check if this PR fixed the problem: https://github.com/apache/tvm/pull/13203

**Comment by MasterJH5574** [comment]
@zxybazh I suppose this PR https://github.com/apache/tvm/pull/13203 should be able to fix this problem. Thanks!

**Comment by zxybazh** [comment]
Fixed by #13203
